### PR TITLE
Add distributed execution support to Collab

### PR DESCRIPTION
**Pull request by YuanTingHsieh**
## What

- Adds optional externally launched Collab rank groups configured with `launch_external_process` and a per-site launcher prefix such as `torchrun`.
- Adds a Collab-owned authenticated child-Cell protocol between each site executor and global rank zero.
- Executes `@collab.init`, every `@collab.publish` call, and `@collab.final` on all ranks in identical order.
- Propagates failures from any rank while returning only the global rank-zero result to the federated workflow.
- Relays rank-zero proxy calls through the parent site Cell.
- Documents the SPMD/multi-node contract and includes a runnable PyTorch DDP example.

## Why

Collab should provide the same launcher-prefix multi-GPU and multi-node promise as Client API without depending on `ClientAPIExecutor`, `ExternalProcessBackend`, or the Client API task protocol.

The launcher creates and supervises the rank processes. DDP initialization, device selection, model wrapping, distributed sampling, barriers, and process-group teardown remain explicit user code in `@collab.init`, `@collab.publish`, and `@collab.final`.

## Implementation notes

- The site executor starts the configured launcher and appends `-m nvflare.collab.runtime.distributed_worker`.
- Global rank zero owns the child Cell connection; distributed collectives broadcast calls and gather rank outcomes.
- Worker bootstrap data is session-scoped, authenticated, owner-only, and written atomically.
- Transport failures are latched so subsequent calls fail consistently instead of re-entering a broken distributed session.
- Finalization uses a fresh teardown signal, waits for all ranks, and then cleans up the launched process group.

## Example

[`examples/advanced/collab/distributed_training`](https://github.com/YuanTingHsieh/NVFlare/tree/feat/collab-ddp-support/examples/advanced/collab/distributed_training) demonstrates federated averaging with one persistent PyTorch DDP rank group per client.

- Uses synthetic regression data; no download is required.
- Defaults to one NVFlare client and two NCCL ranks.
- Uses a per-site `torch.distributed.run` launcher prefix with a unique rendezvous port.
- Initializes NCCL in `@collab.init`.
- Wraps the model with `DistributedDataParallel` and shards data with `DistributedSampler` in `@collab.publish`.
- Returns the trained state only from global rank zero.
- Performs a barrier and destroys the process group in `@collab.final`.
- Includes documentation for multiple federated clients and launcher-owned multi-node execution.

Run from `examples/advanced`:

```bash
python -m collab.distributed_training.distributed_training
```

### Local

- Full style check: `./runtest.sh -s --skip-install`
  - Black: 2,592 files unchanged
  - isort passed
  - flake8: 0
  - agent skill lint: 0 errors and warnings
- Example smoke/config tests: `5 passed, 5 skipped`; PyTorch-dependent rows skip because this local environment has no PyTorch.
- Collab suite: `103 passed, 6 skipped, 1 deselected`.
  - The deselected clean-process import check cannot find `msgpack` in this local base `python -I` environment.
- Unchanged Client API suites: `146 passed, 2 skipped`.

### Colossus runtime qualification

The Collab-owned runtime snapshot in signed commit `e41be5745bd81b71f84166cfa72599611d624f6e` passed the focused distributed/recipe suite (`20 passed`), the full Collab suite (`108 passed`), and a failure/recovery E2E on 2×A40 with real NCCL collectives.

- Both ranks initialized and ran ordered calls.
- Two real DDP collective checksums matched on both ranks.
- An intentional rank-one exception propagated; the following invocation succeeded.
- Only rank zero returned results and issued outbound proxy calls.
- Both ranks completed the final barrier and destroyed the process group.
- No Client API backend/protocol signatures or process/GPU leaks were found.

### Public example qualification

The example was independently run on exact clean PR HEAD `6d61e5c04fb01086e1efca182ab3d00b96148636` (tree `d01bf0269357b434c7de3fc3efd208259f56387a`) on the existing 2×A40 lease. No source edits were made.

- Example smoke/config suite: `10 passed`.
- Exact documented module entry point, with `CUDA_VISIBLE_DEVICES=0,1`: exited 0.
- Real NCCL `world_size=2`; rank zero used device 0 and rank one used device 1.
- All three federated rounds completed; loss decreased `1.7883 → 0.3762 → 0.0863`.
- Both ranks executed six `all_reduce` calls, completed the final barrier, destroyed the process group, and ended with `initialized=false`.
- The server received exactly one valid rank-zero result per round.
- Runtime scan found no `ClientAPIExecutor`, `ExternalProcessBackend`, `CellClientAPI`, `client_api.*`, `TASK_READY`, or `RESULT_READY` signatures.
- No launcher, worker, example, or GPU compute processes remained; both GPUs returned to idle.

### CODE REVIEW

File: nvflare/collab/runtime/distributed_worker.py [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+
+    def _handle_finalize(self, request: Message) -> Message:
+        _, error = self._validate_parent_request(request)
+        if error:
+            return make_reply(ReturnCode.INVALID_REQUEST, error=error)
+        with self.call_lock:
+            if self.finalized:
+                return make_reply(ReturnCode.OK)
+            self._broadcast(DistributedTopic.FINALIZE)
+            return self._finalize_local()
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/collab/runtime/distributed_worker.py:284]: Addressed in `38788c1b2`. A nonzero-rank `@collab.final` failure still exits that worker with a nonzero status, and `DistributedClientSession.stop()` now waits for launcher termination after cleanup, checks the launcher return code, logs the teardown failure, and raises it through `CollabExecutor._handle_end_run()`. This preserves application-owned `destroy_process_group()` while ensuring the parent cannot report successful teardown when another rank failed. Added `test_session_reports_nonzero_worker_exit_after_finalization` and distributed END_RUN cleanup/propagation coverage; the full Collab suite passes (117 tests).

File: nvflare/collab/runtime/distributed_worker.py [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+        timer.daemon = True
+        timer.start()
+        return make_reply(ReturnCode.OK)
+
+    def _build_components(self, bootstrap):
+        with open(bootstrap["client_config"], "r", encoding="utf-8") as f:
+            app_config = json.load(f)
+        entries = {entry.get("id"): (i, entry) for i, entry in enumerate(app_config.get("components", []), 1)}
+        required_ids = [bootstrap["client_obj_id"]] + list(bootstrap.get("collab_obj_ids") or [])
+        builder = WorkerComponentBuilder(workspace=self.workspace, enforce_authorization=False)
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/collab/runtime/distributed_worker.py]: Addressed in `38788c1b2`. Removed `enforce_authorization=False`. The worker now reconstructs an `FLContext` with the actual workspace and job metadata before component building, then uses the default `WorkerComponentBuilder(fl_ctx=..., workspace=...)`. BYOC semantics are therefore preserved, while non-BYOC jobs retain `ComponentPathAuthorizer` allow-list enforcement. Added a regression test that asserts authorization is not disabled.

File: nvflare/collab/runtime/distributed_worker.py [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+                traceback_text=traceback_text,
+            )
+        finally:
+            self.finalized = True
+
+    def _run_nonzero_rank(self):
+        while True:
+            command = [None]
+            self.dist.broadcast_object_list(command, src=0)
+            command = command[0]
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/collab/runtime/distributed_worker.py:498]: Addressed in `38788c1b2`. `self.initialized` is now derived from the collective initialization status, and `_run_nonzero_rank()` returns immediately when initialization failed or `self.dist` is unavailable. Nonzero ranks therefore exit cleanly instead of dereferencing `None`. Added `test_nonzero_rank_skips_broadcast_loop_after_failed_initialization`; the full Collab suite passes (117 tests).
